### PR TITLE
Fix IAM token expiration

### LINKED ISSUE #233: IAM auth: session token eventually expires
Hey!

I noticed an issue in the IAM auth feature I made a PR for a while back. Real world kartons are supposed to be long-running services and shouldn't crash, so essentially `KartonBackend` should be initialized once in their startup process. This means that the s3 client will use the same `session_token`  and eventually it will expire and cause the karton to fail processing tasks 100% of the time.

There are a couple of ways to deal with it I think are worth discussion:  
* Make s3 client on each request: In mwdb, when a file object's `get_or_create` is called, [a new temporary client is made every time](https://github.com/CERT-Polska/mwdb-core/blob/a279a92b99f7787bffdfac804104ef69167470ba/mwdb/model/file.py#L134-L145). This solves the problem because the session token doesn't have time to expire. Similarly, we can move the s3 client creation to a separate method:  
```python3
class KartonBackend:
    def __init__(
        self,
        config: Config,
        identity: Optional[str] = None,
        service_info: Optional[KartonServiceInfo] = None,
    ) -> None:
        self.config = config

        if identity is not None:
            self._validate_identity(identity)
        self.identity = identity

        self.service_info = service_info
        self.redis = self.make_redis(
            config, identity=identity, service_info=service_info
        )

        # don't use this
        # self.s3 = ...

    def make_s3(self):
        session_token = None
        endpoint = self.config.get("s3", "address")
        access_key = self.config.get("s3", "access_key")
        secret_key = self.config.get("s3", "secret_key")
        iam_auth = self.config.getboolean("s3", "iam_auth")

        if not endpoint:
            raise RuntimeError("Attempting to get S3 client without an endpoint set")
        ...

	 def get_object(self, bucket: str, object_uid: str) -> HTTPResponse:
        """
        Get resource object stream with the content.

        Returned response should be closed after use to release network resources.
        To reuse the connection, it's required to call `response.release_conn()`
        explicitly.

        :param bucket: Bucket name
        :param object_uid: Object identifier
        :return: Response object with content
        """
        return self.make_s3().get_object(Bucket=bucket, Key=object_uid)["Body"]
```
However, this can potentially create overhead and reduce the performance because of the constant creation of the client.

* Implement credential refresh mechanism: Tools to do this should already exist in botocore, but this is more complicated overall. One thing that comes to mind is to use botocore's existing codebase to check if credentials need a refresh before every call to `process` method of any karton. So essentially:
```python3
# where the process() method of a karton is
# dispatched. Probably in karton system?
# karton_obj is some karton class instance
if check_token_refresh(karton_obj):
    refresh_token(karton_obj)
karton_obj.process(task)
```

### PR DESCRIPTION
**Pull request by yankovs**
This PR solves the issue described in #233. Botocore's `*Provider` classes such as `ContainerProvider` return a `RefreshableCredentials` object when calling their `load` method. This object knows when and how to refresh the session token, so there's no need to implement custom code to do that. In current code, this object is thrown away so it gets no chance to auto-refresh the token.

Instead, use a lower level session object and if possible, attach this `RefreshableCredentials` object to it, and create the s3 client from it. 

Tested on MWDB v2.10.2 with a karton that uses v5.3.0 of karton-core with the diff in this PR patched into it and seems to still work even a couple of hours after the karton is up (whereas the current code fails). Kept as draft until I'm sure it solves the issue

Fixes #233

### CODE REVIEW

**Comment by yankovs** [comment]
It's really annoying that my solution has to access internal members of botocore classes (hence the `# type: ignore`s), but seems like this is the way to go right now: https://github.com/boto/botocore/pull/1742#issuecomment-1574235418

**Comment by msm-code** [comment]
Hi! Looks good I think.

What do you think about splitting the IAM and "regular" logic? Right now it's a bit tangled. I'm thinking of something like:

```python
def init_snippet():
    # first part of the __init__ method
    if iam_auth:
        self.s3 = iam_auth()
        return  # or `if self.s3 is None: return` to try fallback, like the current code does

    if access_key is None or secret_key is None:
        raise RuntimeError(
                "Attempting to get S3 client without an access_key/secret_key set"
            )
    self.s3 = boto3.Session().client(
        "s3",
        endpoint_url=endpoint,
        aws_access_key_id=access_key,
        aws_secret_access_key=secret_key,
    )

def iam_auth():
    boto_session = get_session()
    iam_providers = [
        ContainerProvider(),
        InstanceMetadataProvider(
            iam_role_fetcher=InstanceMetadataFetcher(
                timeout=1000, num_attempts=2
            )
        ),
    ]
    for provider in iam_providers:
        creds = provider.load()
        if creds:
            boto_session._credentials = creds  # type: ignore
            return boto3.Session(botocore_session=boto_session).client(
                "s3",
                endpoint_url=endpoint,
            )
```
(not tested ofc, but I think it's significantly easier to understand)

**Comment by psrok1** [comment]
Btw, similar code is used in MWDB, so when we agree on final solution I'll be happy to apply the change to mwdb-core as well: https://github.com/CERT-Polska/mwdb-core/blob/master/mwdb/core/util.py#L164

**Comment by yankovs** [comment]
> Hi! Looks good I think.
> 
> What do you think about splitting the IAM and "regular" logic? Right now it's a bit tangled. I'm thinking of something like:
> 
> ```python
> def init_snippet():
>     # first part of the __init__ method
>     if iam_auth:
>         self.s3 = iam_auth()
>         return  # or `if self.s3 is not None: return` to try fallback, like the current code does
> 
>     if access_key is None or secret_key is None:
>         raise RuntimeError(
>                 "Attempting to get S3 client without an access_key/secret_key set"
>             )
>     self.s3 = boto3.Session().client(
>         "s3",
>         endpoint_url=endpoint,
>         aws_access_key_id=access_key,
>         aws_secret_access_key=secret_key,
>     )
> 
> def iam_auth():
>     boto_session = get_session()
>     iam_providers = [
>         ContainerProvider(),
>         InstanceMetadataProvider(
>             iam_role_fetcher=InstanceMetadataFetcher(
>                 timeout=1000, num_attempts=2
>             )
>         ),
>     ]
>     for provider in iam_providers:
>         creds = provider.load()
>         if creds:
>             boto_session._credentials = creds  # type: ignore
>             return boto3.Session(botocore_session=boto_session).client(
>                 "s3",
>                 endpoint_url=endpoint,
>             )
> ```
> 
> (not tested ofc, but I think it's significantly easier to understand)

Sure, if you guys think this is cleaner I have no real objection :)

I do want to note however, that during some reading in the botocore repo I noticed they have some function called [`create_credential_resolver`](https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L126) and inside it is all of the different providers they have are created - not only `ContainerProvider` etc.

So instead of selecting a few providers (in our case 2), we can call `create_credential_resolver` and let it search for credentials in all places they support: env variables, config in `~/.boto` file, etc. This will make `iam_auth` meaningless however. Instead, maybe `require_hardcoded_keys` or something like that will make more sense.

File: karton/core/backend.py [hunk 1]
```diff
@@ -163,9 +151,26 @@ def __init__(
             endpoint_url=endpoint,
             aws_access_key_id=access_key,
             aws_secret_access_key=secret_key,
-            aws_session_token=session_token,
         )
 
+    def iam_auth_s3(self, endpoint: str):
```
Comment (nazywam) [inline karton/core/backend.py:156]:
```suggestion
    def iam_auth_s3(self, endpoint: str) -> Optional[botocore.client.S3]:
```

**Review by nazywam — APPROVED** [review]

**Review by nazywam — APPROVED** [review]

**Pull request merged**